src/ray/gcs/gcs_server/test/gcs_based_actor_scheduler_test.cc

```diff
@@ -0,0 +1,607 @@
+            io_service_, *gcs_actor_table_, *gcs_node_manager_, gcs_pub_sub_,
+            gcs_resource_manager_, resource_scheduler,
+            /*schedule_failure_handler=*/
+            [this](std::shared_ptr<gcs::GcsActor> actor) {
+              failure_actors_.emplace_back(std::move(actor));
+            },
+            /*schedule_success_handler=*/
+            [this](std::shared_ptr<gcs::GcsActor> actor) {
+              success_actors_.emplace_back(std::move(actor));
+            },
+            raylet_client_pool_,
+            /*client_factory=*/
+            [this](const rpc::Address &address) { return worker_client_; });
+  }
+
+  std::shared_ptr<gcs::GcsActor> NewGcsActor(
+      const std::unordered_map<std::string, double> &required_placement_resources =
+          std::unordered_map<std::string, double>()) {
+    rpc::Address owner_address;
+    owner_address.set_raylet_id(NodeID::FromRandom().Binary());
+    owner_address.set_ip_address("127.0.0.1");
+    owner_address.set_port(5678);
+    owner_address.set_worker_id(WorkerID::FromRandom().Binary());
+    auto job_id = JobID::FromInt(1);
+
+    std::unordered_map<std::string, double> required_resources;
+    auto actor_creating_task_spec = Mocker::GenActorCreationTask(
+        job_id, /*max_restarts=*/1, /*detached=*/true, /*name=*/"", owner_address,
+        required_resources, required_placement_resources);
+    return std::make_shared<gcs::GcsActor>(actor_creating_task_spec.GetMessage(),
+                                           /*ray_namespace=*/"");
+  }
+
+  std::shared_ptr<rpc::GcsNodeInfo> AddNewNode(
+      std::unordered_map<std::string, double> node_resources) {
+    auto node_info = Mocker::GenNodeInfo();
+    node_info->mutable_resources_total()->insert(node_resources.begin(),
+                                                 node_resources.end());
+    gcs_node_manager_->AddNode(node_info);
+    gcs_resource_manager_->OnNodeAdd(*node_info);
+    return node_info;
+  }
+
+ protected:
+  instrumented_io_context io_service_;
+  std::shared_ptr<gcs::StoreClient> store_client_;
+  std::shared_ptr<GcsServerMocker::MockedGcsActorTable> gcs_actor_table_;
+  std::shared_ptr<GcsServerMocker::MockRayletClient> raylet_client_;
+  std::shared_ptr<GcsServerMocker::MockWorkerClient> worker_client_;
+  std::shared_ptr<gcs::GcsNodeManager> gcs_node_manager_;
+  std::shared_ptr<gcs::GcsResourceManager> gcs_resource_manager_;
+  std::shared_ptr<GcsServerMocker::MockedGcsBasedActorScheduler> gcs_actor_scheduler_;
+  std::vector<std::shared_ptr<gcs::GcsActor>> success_actors_;
+  std::vector<std::shared_ptr<gcs::GcsActor>> failure_actors_;
+  std::shared_ptr<GcsServerMocker::MockGcsPubSub> gcs_pub_sub_;
+  std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage_;
+  std::shared_ptr<gcs::RedisClient> redis_client_;
+  std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool_;
+};
+
+TEST_F(GcsBasedActorSchedulerTest, TestScheduleFailedWithZeroNode) {
```

Comment:
this is awesome!
![image](https://user-images.githubusercontent.com/837180/128992087-fd768280-562b-40bc-b06c-d75d5068fc89.png)